framework/include/gpu/base/GPUVector.h

```diff
@@ -0,0 +1,157 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GPUArray.h"
+
+#include "libmesh/petsc_vector.h"
+#include "libmesh/dof_map.h"
+
+class GPUSystem;
+
+class GPUVector
+{
+private:
+  struct MPIBuffer
+  {
+    // List of DOFs to send/receive for each process
+    GPUArray<GPUArray<libMesh::dof_id_type>> list;
+    // Number of DOFs to send/receive for each process
+    GPUArray<int> count;
+    // Offset of each process in the buffer
+    GPUArray<int> offset;
+    // Send/receive buffer
+    GPUArray<PetscScalar> buffer;
+    // Allocate data
+    void create(const GPUArray<GPUArray<libMesh::dof_id_type>> & list);
+    // Free data
+    void destroy();
+  };
+
+private:
+  // The global PETSc vector
+  Vec _vector = PETSC_NULLPTR;
+  // The local PETSc vector
+  Vec _local_vector = PETSC_NULLPTR;
+  // The raw array of PETSc vector
+  PetscScalar * _array = PETSC_NULLPTR;
+  // Pointer to the GPU system
+  const GPUSystem * _system;
+  // Pointer to the libMesh communicator
+  const libMesh::Parallel::Communicator * _comm = nullptr;
+  // Send buffer
+  MPIBuffer _send;
+  // Receive buffer
+  MPIBuffer _recv;
+  // Data that is used locally
+  GPUArray<PetscScalar> _local;
+  // Data that should be sent to ghost processes
+  GPUArray<PetscScalar> _ghost;
+  // Flag whether the vector will be assembled
+  bool _assemble = false;
+  // Flag whether the vector is ghosted
+  bool _is_ghosted = false;
+  // Flag whether the PETSc vector is a host vector
+  bool _is_host = false;
+  // Flag whether the vector was allocated
+  bool _is_alloc = false;
```

Comment:
Defaults here also error prone. For example, you have a `bool assemble = false` default in create, but also a default here. It makes it difficult to see where these are actually set. You could construct this object, look at that construction, and not know what's actually going on